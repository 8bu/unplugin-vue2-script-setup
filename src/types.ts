export interface TagMeta {
  start: number
  end: number
  contentStart: number
  contentEnd: number
  content: string
  attrs: Record<string, string>
  found: boolean
}

export interface ParseResult {
  id?: string
  template: {
    components: Set<string>
    identifiers: Set<string>
  }
  scriptSetup: TagMeta
  script: TagMeta
}

export interface PropTypeData {
  key: string
  type: string[]
  required: boolean
}
