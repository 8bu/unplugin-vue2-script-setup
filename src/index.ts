export * from './core/transform'
export * from './types'
