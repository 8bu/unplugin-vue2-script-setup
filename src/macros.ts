import {
  Node,
  Declaration,
  ObjectPattern,
  ObjectExpression,
  ArrayPattern,
  Identifier,
  ExportSpecifier,
  Function as FunctionNode,
  TSType,
  TSTypeLiteral,
  TSFunctionType,
  ObjectProperty,
  ArrayExpression,
  Statement,
  CallExpression,
  RestElement,
  TSInterfaceBody,
  AwaitExpression,
  VariableDeclarator,
  VariableDeclaration,
} from '@babel/types'
import { types as t } from '@babel/core'
import { PropTypeData } from './types'

// Special compiler macros
const DEFINE_PROPS = 'defineProps'
const DEFINE_EMITS = 'defineEmits'
const DEFINE_EXPOSE = 'defineExpose'
const WITH_DEFAULTS = 'withDefaults'

export function applyMacros(nodes: Statement[]) {
  let hasDefinePropsCall = false
  let hasDefineEmitCall = false
  const hasDefineExposeCall = false
  let propsRuntimeDecl: Node | undefined
  let propsRuntimeDefaults: Node | undefined
  let propsTypeDecl: TSTypeLiteral | TSInterfaceBody | undefined
  let propsTypeDeclRaw: Node | undefined
  let propsIdentifier: string | undefined
  let emitsRuntimeDecl: Node | undefined
  let emitsTypeDecl:
  | TSFunctionType
  | TSTypeLiteral
  | TSInterfaceBody
  | undefined
  let emitsTypeDeclRaw: Node | undefined
  let emitIdentifier: string | undefined

  function error(
    msg: string,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    node: Node,
  ): never {
    throw new Error(msg)
  }

  function processDefineProps(node: Node): boolean {
    if (!isCallOf(node, DEFINE_PROPS))
      return false

    if (hasDefinePropsCall)
      error(`duplicate ${DEFINE_PROPS}() call`, node)

    hasDefinePropsCall = true

    propsRuntimeDecl = node.arguments[0]

    // call has type parameters - infer runtime types from it
    if (node.typeParameters) {
      if (propsRuntimeDecl) {
        error(
          `${DEFINE_PROPS}() cannot accept both type and non-type arguments `
            + 'at the same time. Use one or the other.',
          node,
        )
      }

      propsTypeDeclRaw = node.typeParameters.params[0]
      propsTypeDecl = resolveQualifiedType(
        propsTypeDeclRaw,
        node => node.type === 'TSTypeLiteral',
      ) as TSTypeLiteral | TSInterfaceBody | undefined

      if (!propsTypeDecl) {
        error(
          `type argument passed to ${DEFINE_PROPS}() must be a literal type, `
            + 'or a reference to an interface or literal type.',
          propsTypeDeclRaw,
        )
      }
    }

    return true
  }

  function processWithDefaults(node: Node): boolean {
    if (!isCallOf(node, WITH_DEFAULTS))
      return false

    if (processDefineProps(node.arguments[0])) {
      if (propsRuntimeDecl) {
        error(
          `${WITH_DEFAULTS} can only be used with type-based `
            + `${DEFINE_PROPS} declaration.`,
          node,
        )
      }
      propsRuntimeDefaults = node.arguments[1]
    }
    else {
      error(
        `${WITH_DEFAULTS}' first argument must be a ${DEFINE_PROPS} call.`,
        node.arguments[0] || node,
      )
    }
    return true
  }

  function processDefineEmits(node: Node): boolean {
    if (!isCallOf(node, DEFINE_EMITS))
      return false

    if (hasDefineEmitCall)
      error(`duplicate ${DEFINE_EMITS}() call`, node)

    hasDefineEmitCall = true
    emitsRuntimeDecl = node.arguments[0]
    if (node.typeParameters) {
      if (emitsRuntimeDecl) {
        error(
          `${DEFINE_EMITS}() cannot accept both type and non-type arguments `
            + 'at the same time. Use one or the other.',
          node,
        )
      }

      emitsTypeDeclRaw = node.typeParameters.params[0]
      emitsTypeDecl = resolveQualifiedType(
        emitsTypeDeclRaw,
        node => node.type === 'TSFunctionType' || node.type === 'TSTypeLiteral',
      ) as TSFunctionType | TSTypeLiteral | TSInterfaceBody | undefined

      if (!emitsTypeDecl) {
        error(
          `type argument passed to ${DEFINE_EMITS}() must be a function type, `
            + 'a literal type with call signatures, or a reference to the above types.',
          emitsTypeDeclRaw,
        )
      }
    }
    return true
  }

  function resolveQualifiedType(
    node: Node,
    qualifier: (node: Node) => boolean,
  ) {
    if (qualifier(node))
      return node

    if (
      node.type === 'TSTypeReference'
      && node.typeName.type === 'Identifier'
    ) {
      const refName = node.typeName.name
      const isQualifiedType = (node: Node): Node | undefined => {
        if (
          node.type === 'TSInterfaceDeclaration'
          && node.id.name === refName
        )
          return node.body

        else if (
          node.type === 'TSTypeAliasDeclaration'
          && node.id.name === refName
          && qualifier(node.typeAnnotation)
        )
          return node.typeAnnotation

        else if (node.type === 'ExportNamedDeclaration' && node.declaration)
          return isQualifiedType(node.declaration)
      }

      for (const node of nodes) {
        const qualified = isQualifiedType(node)
        if (qualified)
          return qualified
      }
    }
  }

  function processDefineExpose(node: Node): boolean {
    if (isCallOf(node, DEFINE_EXPOSE))
      error(`Vue 2 does not support ${DEFINE_EXPOSE}()`, node)
    return false
  }

  /* function genRuntimeProps(props: Record<string, PropTypeData>) {
    const keys = Object.keys(props)
    if (!keys.length)
      return ''

    // check defaults. If the default object is an object literal with only
    // static properties, we can directly generate more optimzied default
    // decalrations. Otherwise we will have to fallback to runtime merging.
    const hasStaticDefaults
      = propsRuntimeDefaults
      && propsRuntimeDefaults.type === 'ObjectExpression'
      && propsRuntimeDefaults.properties.every(
        node => node.type === 'ObjectProperty' && !node.computed,
      )

    let propsDecls = `{
    ${keys
    .map((key) => {
      let defaultString: string | undefined
      if (hasStaticDefaults) {
        const prop = (
          propsRuntimeDefaults as ObjectExpression
        ).properties.find(
          (node: any) => node.key.name === key,
        ) as ObjectProperty
        if (prop) {
          // prop has corresponding static default value
          defaultString = `default: ${source.slice(
            prop.value.start! + startOffset,
            prop.value.end! + startOffset,
          )}`
        }
      }

      const { type, required } = props[key]
      return `${key}: { type: ${toRuntimeTypeString(
        type,
      )}, required: ${required}${
        defaultString ? `, ${defaultString}` : ''
      } }`
    })
    .join(',\n    ')}\n  }`

    if (propsRuntimeDefaults && !hasStaticDefaults) {
      propsDecls = `${helper('mergeDefaults')}(${propsDecls}, ${source.slice(
        propsRuntimeDefaults.start! + startOffset,
        propsRuntimeDefaults.end! + startOffset,
      )})`
    }

    return `\n  props: ${propsDecls} as unknown as undefined,`
  } */

  nodes = nodes
    .map((node) => {
      if (node.type === 'VariableDeclaration' && !node.declare) {
        const total = node.declarations.length
        for (let i = 0; i < total; i++) {
          const decl = node.declarations[i]
          if (decl.init) {
            if (processDefineEmits(decl.init))
              decl.init = t.memberExpression(t.identifier('__ctx'), t.identifier('emit')) as any
            else if (processDefineProps(decl.init) || processWithDefaults(decl.init))
              decl.init = t.identifier('__props') as any
          }
        }
      }

      if (processDefineEmits(node) || processDefineProps(node) || processDefineExpose(node))
        return null

      return node
    })
    .filter(Boolean) as Statement[]

  return {
    nodes,
    props: propsRuntimeDecl,
  }
}

function isCallOf(
  node: Node | null | undefined,
  test: string | ((id: string) => boolean),
): node is CallExpression {
  return !!(
    node
    && node.type === 'CallExpression'
    && node.callee.type === 'Identifier'
    && (typeof test === 'string'
      ? node.callee.name === test
      : test(node.callee.name))
  )
}
