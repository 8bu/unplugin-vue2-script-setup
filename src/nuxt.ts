import vitePlugin from './vite-plugin'
import webpackPlugin from './webpack-plugin'

export default function(this: any) {
  // install webpack plugin
  this.extendBuild((config: any) => {
    config.plugins = config.plugins || []
    config.plugins.unshift(webpackPlugin())
  })

  // install vite plugin
  this.nuxt.hook('vite:extend', async(vite: any) => {
    vite.config.plugins = vite.config.plugins || []
    vite.config.plugins.push(vitePlugin())
  })
}
