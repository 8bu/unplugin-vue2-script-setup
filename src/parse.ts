import { Parser as HTMLParser } from 'htmlparser2'
import { parse, ParserOptions, ParserPlugin } from '@babel/parser'
import { isHTMLTag, isSVGTag, isVoidTag } from '@vue/shared'
import traverse from '@babel/traverse'
import MagicString from 'magic-string'

interface TagMeta {
  start: number
  end: number
  contentStart: number
  contentEnd: number
  content: string
  attributes: Record<string, string>
}

export interface ParseResult {
  template: {
    components: Set<string>
    identifiers: Set<string>
  }
  scriptSetup: TagMeta
  script: TagMeta
}

export function parseVueSFC(code: string): ParseResult {
  const components = new Set<string>()
  const expressions = new Set<string>()
  const identifiers = new Set<string>()

  let templateLevel = 0
  let inScriptSetup = false
  let inScript = false

  const scriptSetup: TagMeta = {
    start: 0,
    end: 0,
    contentStart: 0,
    contentEnd: 0,
    content: '',
    attributes: {},
  }
  const script: TagMeta = {
    start: 0,
    end: 0,
    contentStart: 0,
    contentEnd: 0,
    content: '',
    attributes: {},
  }

  const parser = new HTMLParser({
    onopentag(name, attributes) {
      if (name === 'template')
        templateLevel += 1

      if (templateLevel > 0) {
        if (!isHTMLTag(name) && !isSVGTag(name) && !isVoidTag(name))
          components.add(name)
        Object.entries(attributes).forEach(([key, value]) => {
          if (!value)
            return
          if (key.startsWith('v-') || key.startsWith('@') || key.startsWith(':'))
            expressions.add(value)
          if (key === 'ref')
            identifiers.add(value)
        })
      }
      else {
        if (name === 'script') {
          if ('setup' in attributes) {
            scriptSetup.start = parser.startIndex
            scriptSetup.contentStart = parser.endIndex! + 1
            scriptSetup.attributes = attributes
            inScriptSetup = true
          }
          else {
            script.start = parser.startIndex
            script.contentStart = parser.endIndex! + 1
            script.attributes = attributes
            inScript = true
          }
        }
      }
    },
    ontext(text) {
      if (templateLevel > 0) {
        Array.from(text.matchAll(/\{\{(.*?)\}\}/g)).forEach(([, expression]) => {
          expressions.add(expression)
        })
      }
    },
    onclosetag(name) {
      if (name === 'template')
        templateLevel -= 1

      if (inScriptSetup && name === 'script') {
        scriptSetup.end = parser.endIndex!
        scriptSetup.contentEnd = parser.startIndex
        scriptSetup.content = code.slice(scriptSetup.contentStart, scriptSetup.contentEnd)
        inScriptSetup = false
      }
      if (inScript && name === 'script') {
        script.end = parser.endIndex!
        script.contentEnd = parser.startIndex
        script.content = code.slice(script.contentStart, script.contentEnd)
        inScript = false
      }
    },
  }, {
    xmlMode: true,
    lowerCaseTags: false,
    lowerCaseAttributeNames: false,
    recognizeSelfClosing: true,
  })

  parser.write(code)
  parser.end()

  expressions.forEach(exp => getIdentifiersFromCode(exp, identifiers))

  return {
    template: {
      components,
      identifiers,
    },
    scriptSetup,
    script,
  }
}

export function getIdentifiersFromCode(code: string, identifiers = new Set<string>(), options: ParserOptions = {}) {
  const ast = parse(code, options) as any
  traverse(ast, {
    Identifier(path) {
      identifiers.add(path.node.name)
    },
  })
  return identifiers
}

export function parseScriptSetup(result: ParseResult) {
  const code = result.scriptSetup.content
  const s = new MagicString(code)

  const plugins: ParserPlugin[] = []
  if (result.scriptSetup.attributes.lang === 'ts')
    plugins.push('typescript')
  if (result.scriptSetup.attributes.lang === 'jsx')
    plugins.push('jsx')
  if (result.scriptSetup.attributes.lang === 'tsx')
    plugins.push('typescript', 'jsx')

  const identifiers = new Set<string>()
  const ast = parse(code, {
    sourceType: 'module',
    plugins,
  })

  const imports: string[] = []

  traverse(ast as any, {
    Identifier(path) {
      identifiers.add(path.node.name)
    },
    ImportDeclaration(path) {
      imports.push(code.slice(path.node.start!, path.node.end!))
      s.remove(path.node.start!, path.node.end!)
    },
  })

  const returns = Array.from(identifiers).filter(i => result.template.identifiers.has(i))

  s.prepend('() => {')
  s.append(`\nreturn { ${returns.join(',')} }\n}`)

  return {
    imports,
    returns,
    ast,
    fn: s.toString(),
  }
}
