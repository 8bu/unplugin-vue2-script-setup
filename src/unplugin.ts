import { createUnplugin } from 'unplugin'
import { ScriptSetupTransformOptions } from './types'
import { transform } from '.'

const scriptSetupRE = /<script\s+setup/

export default createUnplugin<ScriptSetupTransformOptions>(options => ({
  name: 'vue2-script-setup-transform',
  enforce: 'pre',
  transformInclude(id) {
    return id.endsWith('.vue')
  },
  transform(code, id) {
    try {
      if (scriptSetupRE.test(code))
        return transform(code, id, options)
    }
    catch (e) {
      this.error(e)
    }
  },
}))
