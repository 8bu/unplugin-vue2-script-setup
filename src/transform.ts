import MagicString from 'magic-string'
import { parseScriptSetup, parseVueSFC } from './parse'

export function transform(sfc: string) {
  const s = new MagicString(sfc)
  const result = parseVueSFC(sfc)
  const script = parseScriptSetup(result)

  const full = [
    '',
    ...script.imports,
    'const __sfc_main = {}',
    `__sfc_main.setup = ${script.fn}`,
    'export default __sfc_main',
    '',
  ].join('\n')

  s.overwrite(
    result.scriptSetup.contentStart,
    result.scriptSetup.contentEnd,
    full,
  )

  return s.toString()
}
