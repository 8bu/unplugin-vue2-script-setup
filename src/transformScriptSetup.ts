import { types as t } from '@babel/core'
import { parse, ParserPlugin } from '@babel/parser'
import { camelize, capitalize } from '@vue/shared'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import { ParseResult } from './types'
import { applyMacros } from './macros'
import { getIdentifiersDeclaration } from './parse'

export function transformScriptSetup(result: ParseResult) {
  if (result.script.found && result.scriptSetup.found && result.scriptSetup.attrs.lang !== result.script.attrs.lang)
    throw new SyntaxError('<script setup> language must be the same as <script>')

  const lang = result.scriptSetup.attrs.lang || result.script.attrs.lang || 'js'
  const plugins: ParserPlugin[] = []
  if (lang === 'ts')
    plugins.push('typescript')
  else if (lang === 'jsx')
    plugins.push('jsx')
  else if (lang === 'tsx')
    plugins.push('typescript', 'jsx')
  else if (lang !== 'js')
    throw new SyntaxError(`Unsupported script language: ${lang}`)

  const identifiers = new Set<string>()
  const scriptSetupAst = parse(result.scriptSetup.content, {
    sourceType: 'module',
    plugins,
  })
  const scriptAst = parse(result.script.content || '', {
    sourceType: 'module',
    plugins,
  })

  const imports = scriptSetupAst.program.body.filter(n => n.type === 'ImportDeclaration')
  const nodes = scriptSetupAst.program.body.filter(n => n.type !== 'ImportDeclaration')

  const { nodes: scriptSetupBody, props } = applyMacros(nodes)

  // get all identifiers in `<script setup>`
  getIdentifiersDeclaration([...imports, ...nodes], identifiers)

  // filter out identifiers that are used in `<template>`
  const returns = Array.from(identifiers)
    .filter(Boolean)
    .filter(i => result.template.identifiers.has(i))
  const components = Array.from(identifiers)
    .filter(Boolean)
    .filter(i => result.template.components.has(i)
      || result.template.components.has(camelize(i))
      || result.template.components.has(capitalize(camelize(i))),
    )

  // append `<script setup>` imports to `<script>`
  scriptAst.program.body.unshift(...imports)

  const __sfc = t.identifier('__sfc_main')

  let hasBody = false

  // replace `export default` with a temproray variable
  // `const __sfc_main = { ... }`
  traverse(scriptAst as any, {
    ExportDefaultDeclaration(path) {
      hasBody = true
      const decl = path.node.declaration
      path.replaceWith(
        t.variableDeclaration('const', [
          t.variableDeclarator(
            __sfc,
            decl as any,
          ),
        ]),
      )
    },
  })

  // inject `const __sfc_main = {}` if `<script>` has default export
  if (!hasBody) {
    scriptAst.program.body.push(
      t.variableDeclaration('const', [
        t.variableDeclarator(
          __sfc,
          t.objectExpression([]),
        ),
      ]),
    )
  }

  // inject props function
  // `__sfc_main.props = { ... }`
  if (props) {
    hasBody = true
    scriptAst.program.body.push(
      t.expressionStatement(
        t.assignmentExpression('=',
          t.memberExpression(__sfc, t.identifier('props')),
          props as any,
        ),
      ) as any,
    )
  }

  // inject setup function
  // `__sfc_main.setup = () => {}`
  if (nodes.length) {
    hasBody = true
    const returnStatement = t.returnStatement(
      t.objectExpression(
        returns.map((i) => {
          const id = t.identifier(i)
          return t.objectProperty(id, id, false, true)
        }),
      ),
    )

    scriptAst.program.body.push(
      t.expressionStatement(
        t.assignmentExpression('=',
          t.memberExpression(__sfc, t.identifier('setup')),
          t.arrowFunctionExpression([
            t.identifier('__props'),
            t.identifier('__ctx'),
          ], t.blockStatement([
            ...scriptSetupBody,
            returnStatement as any,
          ])),
        ),
      ) as any,
    )
  }

  // inject components
  // `__sfc_main.components = Object.assign({ ... }, __sfc_main.components)`
  if (components.length) {
    hasBody = true
    const componentsObject = t.objectExpression(
      components.map((i) => {
        const id = t.identifier(i)
        return t.objectProperty(id, id, false, true)
      }),
    )

    scriptAst.program.body.push(
      t.expressionStatement(
        t.assignmentExpression('=',
          t.memberExpression(__sfc, t.identifier('components')),
          t.callExpression(
            t.memberExpression(t.identifier('Object'), t.identifier('assign')),
            [
              componentsObject,
              t.memberExpression(__sfc, t.identifier('components')),
            ],
          ),
        ),
      ) as any,
    )
  }

  if (!hasBody) {
    return {
      ast: null,
      code: '',
    }
  }

  // re-export
  // `export default __sfc_main`
  scriptAst.program.body.push(
    t.exportDefaultDeclaration(__sfc) as any,
  )

  return {
    ast: scriptAst,
    code: generate(scriptAst as any).code,
  }
}
