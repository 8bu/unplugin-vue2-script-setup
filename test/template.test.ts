import { parseVueSFC } from '../src'

describe('template', () => {
  it('empty', () => {
    expect(parseVueSFC('').template).toMatchSnapshot()
  })

  it('v-if', () => {
    expect(parseVueSFC('<template><div v-if="foo" v-else /></template>').template).toMatchSnapshot()
  })

  it('v-bind', () => {
    expect(parseVueSFC('<template><div v-bind:value="foo" :value="bar" /></template>').template).toMatchSnapshot()
  })

  it('v-on', () => {
    expect(parseVueSFC('<template><div v-on:click="foo" @click="bar()" /></template>').template).toMatchSnapshot()
  })

  it('ref', () => {
    expect(parseVueSFC('<template><div ref="foo" /></template>').template).toMatchSnapshot()
  })

  it('expressions', () => {
    expect(parseVueSFC('<template><HelloWorld>{{ foo(bar) }} x {{ foobar * 2 }}</HelloWorld></template>').template).toMatchSnapshot()
  })
})
