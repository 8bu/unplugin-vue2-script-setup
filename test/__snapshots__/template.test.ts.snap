// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template empty 1`] = `
Object {
  "components": Set {},
  "identifiers": Set {},
}
`;

exports[`template expressions 1`] = `
Object {
  "components": Set {
    "HelloWorld",
  },
  "identifiers": Set {
    "foo",
    "bar",
    "foobar",
  },
}
`;

exports[`template ref 1`] = `
Object {
  "components": Set {},
  "identifiers": Set {
    "foo",
  },
}
`;

exports[`template v-bind 1`] = `
Object {
  "components": Set {},
  "identifiers": Set {
    "foo",
    "bar",
  },
}
`;

exports[`template v-if 1`] = `
Object {
  "components": Set {},
  "identifiers": Set {
    "foo",
  },
}
`;

exports[`template v-on 1`] = `
Object {
  "components": Set {},
  "identifiers": Set {
    "foo",
    "bar",
  },
}
`;
